File: frontend/src/types/inventory.ts
export interface Inventory {
  id: number;
  product_id: number;

  product_name: string;
  product_category: string | null;

  quantity: number;
  low_stock_threshold: number;

  created_at: string;
  updated_at: string;
}

export interface InventoryUpdateData {
  low_stock_threshold?: number;
}

export type StockMovementType = "OPENING_BALANCE" | "SUPPLIER_RECEIPT" | "MANUAL_INCREASE" | "MANUAL_DECREASE" | "ORDER_DEDUCTION" | "CANCELLATION_RESTORATION";
export interface StockMovement { id: number; inventory_id: number; product_id: number; product_name: string; supplier_id: number | null; supplier_name: string | null; admin_name: string | null; movement_type: StockMovementType; quantity_change: number; quantity_before: number; quantity_after: number; reason: string | null; reference: string | null; source_type: string | null; source_id: number | null; created_at: string; }
export interface StockMovementInput { low_stock_threshold?: number; movement_type: "SUPPLIER_RECEIPT" | "MANUAL_INCREASE" | "MANUAL_DECREASE"; quantity_change: number; reason?: string | null; supplier_id?: number | null; reference?: string | null; }
export interface BatchStockReceiptLine { inventory_id: number; quantity: number; }
export interface BatchStockReceiptInput { supplier_id: number; reference: string; items: BatchStockReceiptLine[]; }
export interface BatchStockReceiptResult { received_count: number; inventories: Inventory[]; }

File: backend/app/schemas/inventory.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class InventoryBase(BaseModel):
    quantity: int = Field(default=0, ge=0)
    low_stock_threshold: int = Field(default=10, ge=0)


class InventoryCreate(InventoryBase):
    product_id: int


class InventoryUpdate(BaseModel):
    quantity: int | None = Field(default=None, ge=0)
    low_stock_threshold: int | None = Field(default=None, ge=0)


class InventoryPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    product_name: str
    product_category: str | None
    quantity: int
    low_stock_threshold: int
    created_at: datetime
    updated_at: datetime

class InventoryNested(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    product_id: int
    quantity: int
    low_stock_threshold: int

class InventoryAdjustment(BaseModel):
    quantity_change: int
    reason: str | None = Field(default=None, max_length=2000)


class StockMovementCreate(BaseModel):
    low_stock_threshold: int | None = Field(default=None, ge=0)
    movement_type: str = Field(pattern="^(SUPPLIER_RECEIPT|MANUAL_INCREASE|MANUAL_DECREASE)$")
    quantity_change: int
    reason: str | None = Field(default=None, max_length=2_000)
    supplier_id: int | None = None
    reference: str | None = Field(default=None, max_length=160)


class BatchStockReceiptLine(BaseModel):
    inventory_id: int = Field(gt=0)
    quantity: int = Field(ge=1)


class BatchStockReceiptCreate(BaseModel):
    supplier_id: int = Field(gt=0)
    reference: str = Field(min_length=2, max_length=160)
    items: list[BatchStockReceiptLine] = Field(min_length=1, max_length=50)


class BatchStockReceiptResult(BaseModel):
    received_count: int
    inventories: list[InventoryPublic]


class OpeningBalanceCreate(BaseModel):
    reason: str = Field(min_length=2, max_length=2_000)


class StockMovementPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    inventory_id: int
    product_id: int
    product_name: str
    supplier_id: int | None
    supplier_name: str | None
    admin_name: str | None
    movement_type: str
    quantity_change: int
    quantity_before: int
    quantity_after: int
    reason: str | None
    reference: str | None
    source_type: str | None
    source_id: int | None
    created_at: datetime
